### PR TITLE
`Orchestrator.RunnableProjects`: value forms refactoring

### PR DESCRIPTION
**Pull request by vlada-shubina**
### Problem
Current implementation of value forms combines factory and instance in same class. This makes logic unclear and further refactoring very difficult

### Solution
Introduced factories for value forms. The factories create actual instances. 
The factories are based on the following base implementations:
- actionable value form - doesn't need any configuration, processes string value to string value
- configurable value form - has configuration, that can be parsed from json; processes string value to string value based on configuration
- dependent value form - has configuration, that can be parsed from json; to process value in addition to configuration needs to know other value forms defined for the template.

### Checks:
- [ ] Added unit tests
- [x] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/TemplateConfigModel.cs [hunk 1]
```diff
@@ -415,9 +415,11 @@ private static BaseSymbol SetupDefaultNameSymbol(string? sourceName)
             Dictionary<string, IValueForm> formMap = new Dictionary<string, IValueForm>(StringComparer.Ordinal);
 
             // setup all the built-in default forms.
-            foreach (KeyValuePair<string, IValueForm> builtInForm in ValueFormRegistry.AllForms)
+            // name of the form is form identifier
+            // this is only possible for the forms that don't need configuration
+            foreach (KeyValuePair<string, IValueFormFactory> builtInForm in ValueFormRegistry.FormLookup)
             {
-                formMap[builtInForm.Key] = builtInForm.Value;
+                formMap[builtInForm.Key] = builtInForm.Value.Create();
```
Comment (vlada-shubina) [inline src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/TemplateConfigModel.cs:422]: this logic is a bit vague and undocumented.
It basically allows to refer to value form based on its identifier; and not the name.
In this case default configuration is used (which doesn't make a lot of sense for configurable forms).
I kept this logic as is, not to introduce breaking change if anyone is using it.

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/RunnableProjectConfig.cs [hunk 2]
```diff
@@ -664,8 +664,11 @@ private IReadOnlyList<IReplacementTokens> ProduceSymbolFilenameReplacements()
                         if (_configuration.Forms.TryGetValue(formName, out IValueForm valueForm))
                         {
                             string symbolName = symbol.Name + "{-VALUE-FORMS-}" + formName;
-                            string processedFileReplacement = valueForm.Process(_configuration.Forms, p.FileRename);
-                            GenerateFileReplacementsForSymbol(processedFileReplacement, symbolName, filenameReplacements);
+                            string? processedFileReplacement = valueForm.Process( p.FileRename, _configuration.Forms);
```
Comment (JanKrivanek) [inline src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/RunnableProjectConfig.cs:667]: nit: space after `(`. Interesting we do not have analyzer rule for this.

**Review by JanKrivanek — APPROVED** [review]
Looks good.

Though I haven't inspected in details the actual factories (except for the new bases - `BaseValueFormFactory` and `DependantValuaFormFactory`) as they should be functionally identical to replaced models

File: src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/RunnableProjectConfig.cs (hunk 2 above)
Comment (vlada-shubina) [reply src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/RunnableProjectConfig.cs:667]: Enabled IDE0055 for this

**Pull request merged**